guidance/choosing-persistence.md

```diff
@@ -0,0 +1,183 @@
+}
+```
+
+At runtime, NHibernate will dynamically create and compile a class that inherits from the saga data class, similar in concept to the following code:
+
+```
+public class  OrderSagaDataProxy : OrderSagaData
+{
+    public override OrderId
+    {
+        get { return base.OrderId; }
+        set { base.OrderId = value; }
+    }
+    
+    public override List<OrderItem> Items
+    {
+        get { return base.Items; }
+        set [ base.Items = value; ]
+    }
+}
+```
+
+Except, the dynamic proxy will be a little more complex because it will include NHibernate magic in each of the intercepted properties, which give it the ability to lazy-load data from the OrderItem table, for example, in order to execute queries against the database only when necessary.
+
+If properties were not marked as `virtual`, NHibernate would be unable to override them in its generated proxy class. So instead, NHibernate persistence will throw an exception at startup time to warn you that non-virtual properties are not allowed.
+
+NOTE: Wherever possible, we endeavor to present all Saga samples and documentation using `virtual` properties, in order to accommodate the NHibernate persister.
+
+#### Saga collections
+
+Because NHibernate must store its data in a relational data store, certain class patterns that don't map well cannot be supported.
+
+```
+public class BatchSagaData : ContainSagaData
+{
+    // Will not work with NHibernate
+    public virtual List<int> ProcessedIds { get; set; }
+}
+```
+
+The NHibernate persister is not able to map `BatchSagaData` to a relational model, as it is not clear what to do with the `List<int>`.
+
+Instead, this concept must be expressed in a different way:
+
+```
+public class BatchSagaData : ContainSagaData
+{
+    public virtual List<BatchId> ProcessedIds { get; set; }
+}
+
+public class BatchId
+{
+    public virtual Guid Id { get; set; }
+    public virtual int ProcessedId { get; set; }
+}
+```
+
+Now, it is clear to the NHibernate persister that it should create two tables, `BatchSagaData` and `BatchId`, and how to relate them together.
+
+#### Saga contention
+
+The NHibernate saga persister takes locks on the saga data table(s) during message processing in order to guarantee that another message processed in parallel does not modify the saga in the meantime. Because of these locks, it's easy for a saga to become a contention point when messages are processed in parallel.
+
+In order to minimize contention, it is important to design sagas such that they are message sending state machines, i.e. they receive a message, make some decisions, and then send/publish other messages or request timeouts. A saga should not query database resources, call web services, or do anything else that will hold the database lock on the saga any longer than absolutely necessary. Instead, the saga should send messages to normal endpoints outside the scope of the saga, which can reply back to the saga with the result if necessary.
```

Comment:
Performance can also be tweaked by using custom mappings where custom mappings contain specific eager or lazy loading patterns on how the data is accessed.